Added a formula evaluator to replace TFormula calls

Dr15Jones opened the pull request:
Created a reco::FormulaEvaluator class which can parse a subset of
the TFormula expressions. The parsing and function evaluation are
done using stateless classes so are thread safe and extremely thread
efficient. The formula parser is shared by all reco::FormulaEvaluators
to save memory and since the parser is stateless it can efficiently
be called by multiple threads simultaneously.

This is a back port from CMSSW_7_6 of #11516

Dr15Jones committed:
Added a formula evaluator to replace TFormula calls

Created a reco::FormulaEvaluator class which can parse a subset of
the TFormula expressions. The parsing and function evaluation are
done using stateless classes so are thread safe and extremely thread
efficient. The formula parser is shared by all reco::FormulaEvaluators
to save memory and since the parser is stateless it can efficiently
be called by multiple threads simultaneously.

cmsbuild commented:
A new Pull Request was created by @Dr15Jones (Chris Jones) for CMSSW_7_5_X.

Added a formula evaluator to replace TFormula calls

It involves the following packages:

CommonTools/Utils


@cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please review it and eventually sign? Thanks.
@makortel this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 6a7bbaaaf69369bbf5e0adf5f7d0f58bc455ba64
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11555/8485/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11555/8485/summary.html

Dr15Jones committed:
Added tests for formulas used by SimpleJetCorrector

Dr15Jones committed:
Added an adaptor for handling arrays

SimpleJetCorrector places its variables and parameters into arrays
instead of std::vector. Adding an adaptor for arrays allows for
direct uses of the array without having to fill a temporary std::vector.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/11555	 6a7bbaa
- the same (all new) code as in #1516, no effect on running workflows
- jenkins tests pass

Dr15Jones committed:
Fixed spelling of precedence

Dr15Jones committed:
Added comparison operators

SimpleJetCorrector uses TFormula expressions which include comparison
operators.

Dr15Jones committed:
Fix precedence error and added additional functions

Further tests of expressions from the data base showed that there
were additional ones that couldn't be handled.

Dr15Jones commented:
please test

Dr15Jones commented:
I've added back ported all the fixes already added to CMSSW_7_6

cmsbuild commented:
The tests are being triggered in jenkins.
JENKINS_TEST_URL

cmsbuild commented:
Pull request #11555 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 304a65389d79cb7cee1276fdbf4c0d2d920720c6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11555/8752/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11555/8752/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/11555	304a653
- with this PR, the parts of CommonTools/Utils related to the formula evaluator are now the same as in 76X
- jenkins tests pass; the code is not used yet in any workflow and comparisons with baseline show no differences.

@Dr15Jones are you going to apply the TFormula change in the JetCorrector in 75X as well?

Dr15Jones commented:
It would be useful for improving threading efficiency but ultimately it is a call for RECO and the L1s. @davidlange6

slava77 commented:
On 10/13/15 9:07 AM, Chris Jones wrote:
> It would be useful for improving threading efficiency but ultimately it
> is a call for RECO and the L1s. @davidlange6
> <https://github.com/davidlange6>

Less change is less cause for possible disruption.
If we don't put it back to 74X, then there is no big need to have it in 75X.
I wouldn't want to have a whole in features in 75X if they get
backported to 74X.



> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11555#issuecomment-147625633>.
>
